[TritonCTS] Updated function to find clock nets

MarceloDanigno opened the pull request:
This is the continuation of PR #280.

The description of the last PR was:

> [PR for Review!!!]
This changes how clocks are loaded into TritonCTS.
First, we create a set and use dbSta's findClkNets function. After that, we insert all clocks from the user inputs (if there are any and if they aren't already in the set.)
After that, we create the Clock objects inside TritonCTS for the nets that have more than 2 sinks. This includes input and inout pins.
 Another thing that was added to this PR was a unit test that test the new functionality. This was based on Cherry's dbSta find_clock unit test. Also, the reports for TritonCTS changed a little, to report cases where pins do not have shapes, the number of clocks, and others...
 [Code in this PR was made with the help/additions of Matt Liberty, James Cherry, and Tom Spyrou]

This PR moves the OpenRoad singleton to the initialization of DbWrapper (the class responsible for finding the clocks in TritonCTS), removes unnecessary commits, adds more error handling in case the clocks do not exist and also updates TritonCTS's reports (changes from  PR#274).

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -74,25 +77,60 @@ void DbWrapper::initDB() {
                 }
                 if (!net) {
                         std::cout << " [WARNING] Net \"" << name << "\" not in design. Skipping...\n";
                         continue;
```
continue is a poor choice here. Use else instead.

MarceloDanigno replied:
Was updated in the last commit.

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -74,25 +77,60 @@ void DbWrapper::initDB() {
+        }
+        for (odb::dbNet* net : clockNets){
+                if (!net) {
+                        odb::dbITerm* iterm = _block->findITerm(net->getName().c_str());
```
This is silly. _openSta->findClkNets returns dbNets and this loop find an dbITerm using the net's **name** and then gets the net back from the iterm. It should not be doing either. It is pointless and ill advised for the nets returned by findClkNets and redundant from the clock name loop (why does that even exist?) also.

MarceloDanigno replied:
This was changed in the last commit.

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -74,25 +77,60 @@ void DbWrapper::initDB() {
+                        net = iterm->getNet();
+                }
+                if (!net) {
+                        std::cout << " [WARNING] Net \"" << net->getName() << "\" not in design. Skipping...\n";
```
If the net is null (which should be tested by != nullptr, not using the boolean operator !), then you can't possibly get the name of the null net to print in the warning.

jjcherry56 commented on src/TritonCTS/test/src/find_clock_pad/Nangate45/Nangate45.lef:
```diff
@@ -0,0 +1,12156 @@
+# 
```
You should NOT be adding another copy of the contents of the Nangate45 directory to the  repo. 
As I pointed out in the monday zoom call, use a symbolic link. From doc/OpenRoadArch.md

The Nangate45 open source library data used by many tests is in
`test/Nangate45`. Use the following command to add a link in the tool
command

```
cd tool/test
ln -s ../../../test/Nangate45
```

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -74,25 +77,60 @@ void DbWrapper::initDB() {
                         odb::dbITerm* iterm = _block->findITerm(name.c_str());
+                        if (!iterm) {
+                                std::cout << " [WARNING] Net \"" << name << "\" not in design. Skipping...\n";
+                                continue;
```
just take continue out of your vocabulary

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -74,25 +77,60 @@ void DbWrapper::initDB() {
+        _openSta->findClkNets(clockNets);
+
+        //Checks the user input in case there are other nets that need to be added to the set.
         std::vector<std::string> clockNetNames;
```
why are there clock net names? where are they coming from?

MarceloDanigno replied:
These are only utilized when the user uses the "-clk_nets" parameter for TritonCTS. That parameter requires a string for the clock nets (the divisor being a space). This is transformed into a vector of strings by parseClockNames and then into each respective net in "odb::dbNet* net = _block->findNet(name.c_str());".

I'm not sure if this parameter is used on jenkins or some benchmark. It is used by some unit tests.

tspyrou replied:
I think the point is that it would be better to fix code to eliminate the need to parse the clock names. In net there is a net->get_name() being passed into a block->find_Net(). This is slow but also may not always work if the naming conventions do not align. If you can modify the code to accept a vector of dbNet* instead of strings it would simplify the code and make it more reliable.

MarceloDanigno commented:
> 
> 
> Can you make sure this works with bp_single. It is not in the CI.

I've tested for bp_single and it seems to work fine. The net that was causing problems is now being found on TritonCTS. However, there are a couple of nets that are found that I'm not sure about:

> Initializing clock net for : "co2_tkn_li"
 Clock net "co2_tkn_li" has 20 sinks
 Initializing clock net for : "co_tkn_li"
 Clock net "co_tkn_li" has 20 sinks
Initializing clock net for : "router_tag_lines_lo_clk_"
 Clock net "router_tag_lines_lo_clk_" has 118 sinks

Are these the right ones? There are a total of 8 clock nets (which should be the right amount).

@jjcherry56 Thanks for the feedback. A couple of topics that you mentioned appear in other places in TritonCTS. For now, I'll focus on removing the occurrences on the files for the current PR.

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -125,7 +163,7 @@ void DbWrapper::initClock(odb::dbNet* net) {
                 odb::dbITerm* iterm = *itr;
 
-                if (iterm->getIoType() != odb::dbIoType::INPUT) {
+                if (!(iterm->isInputSignal())) {
```
This does not look right at all. A clock can be defined on an internal net and the driver is not an input iterm. In the problem this fix is trying to address the clock net that matters is driven by the pad instance.

MarceloDanigno replied:
Here we iterate over all iterms in the net. What you're suggesting is that we should do the same for bterms? Or is there a way to know exactly which cells are connected to a certain net?

jjcherry56 commented:
I suppose it is a style issue but I prefer having the tcl command to do the
arg checking and pass objects to c++. If the clock name is not found it
should call tcl error to throw out of the command and stop reading commands
from a file or return to the command interpreter (it matters).

The preferred way to specify clocks is with sdc. That is one of the reasons
sdc exists. It is fine to have unit tests use a back door instead but I
would not document those options so there is no user confusion about how to
specify clocks to users. It doesn't make sense for every command to have
it's own arguments for them like this.

On Thu, May 28, 2020 at 1:50 PM Marcelo Freire Danigno <
notifications@github.com> wrote:

> *@MarceloDanigno* commented on this pull request.
> ------------------------------
>
> In src/TritonCTS/src/DbWrapper.cpp
> <https://github.com/The-OpenROAD-Project/OpenROAD/pull/287#discussion_r432115342>
> :
>
> > @@ -74,25 +77,60 @@ void DbWrapper::initDB() {
>
>  void DbWrapper::initAllClocks() {
>          std::cout << " Initializing clock nets\n";
> -
> +
> +        clearNumClocks();
> +
> +        std::cout << " Looking for clock nets in the design\n";
> +
> +        //Uses dbSta to find all clock nets in the design.
> +
> +        std::set<odb::dbNet*> clockNets;
> +
> +        _openSta->findClkNets(clockNets);
> +
> +        //Checks the user input in case there are other nets that need to be added to the set.
>          std::vector<std::string> clockNetNames;
>
> These are only utilized when the user uses the "-clk_nets" parameter for
> TritonCTS. That parameter requires a string for the clock nets (the divisor
> being a space). This is transformed into a vector of strings by
> parseClockNames and then into each respective net in "odb::dbNet* net =
> _block->findNet(name.c_str());".
>
> I'm not sure if this parameter is used on jenkins or some benchmark. It is
> used by some unit tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/The-OpenROAD-Project/OpenROAD/pull/287#discussion_r432115342>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJDZFPJ5QYZW36S53JTD5LRT3FBNANCNFSM4NNJ5ITQ>
> .
>

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -125,7 +163,7 @@ void DbWrapper::initClock(odb::dbNet* net) {
         for (itr = iterms.begin(); itr != iterms.end(); ++itr) {
```
range iteration is a lot easier to use and read here.
for (odb::dbITerm *iterm : iterms) 
no ugly c++ iterator necessary in c++11

maliberty replied:
I have a change to this file that eliminates many such usages.  I've been holding it until Eder could merge his PR.

jjcherry56 commented on src/TritonCTS/src/DbWrapper.cpp:
```diff
@@ -189,9 +226,10 @@ void DbWrapper::computeITermPosition(odb::dbITerm* term, DBU &x, DBU &y) const {
-
-        x /= numShapes;
-        y /= numShapes;
+       if (numShapes > 0){
```
Why doesn't this use iterm::getAvgXY instead of grovelling around in the shapes?

MarceloDanigno commented:
I've updated the codes with the changes from most comments. Let me know if the changes to the tests are OK. About the computations on the X and Y positions of sinks, I'll have to make sure with Mateus before changing it, since I didn't review that piece of code yet.

For the clock nets from the user input, I've changed the test to run during the TCL command instead of during TritonCTS. Thus, during clock_tree_synthesis, the user's input would be std::vector<odb::dbNet*> instead of a std::string. If the net isn't in the DB, the TCL command exits, instead of raising an error in C++.

Also, it seems there was an error when building on the jenkins, so I'll try the build again once that clears up.

jjcherry56 commented:
You need to merge the openroad branch to get jenkins to pass because of the fastroute failure.
Same branch. Same PR.
